source/h2_connection.c

```diff
@@ -527,7 +560,178 @@ static void s_try_write_outgoing_frames(struct aws_h2_connection *connection) {
+            }
+        }
+    }
+
+    /* #TODO An endpoint that receives any frame other than PRIORITY after receiving a RST_STREAM
+     * MUST treat that as a stream error (Section 5.4.2) of type STREAM_CLOSED */
+
+    /* Stream was closed long ago, or didn't fit criteria for being ignored */
+    CONNECTION_LOGF(
+        ERROR,
+        connection,
+        "Illegal to receive %s frame on stream id=%" PRIu32 " state=closed",
+        aws_h2_frame_type_to_str(frame_type),
+        stream_id);
+
+    return aws_raise_error(AWS_ERROR_HTTP_STREAM_CLOSED);
+}
+
 /* Decoder callbacks */
+
+int s_decoder_on_headers_begin(uint32_t stream_id, void *userdata) {
+    struct aws_h2_connection *connection = userdata;
+
+    if (connection->base.server_data) {
+        /* Server would create new request-handler stream... */
+        return aws_raise_error(AWS_ERROR_UNIMPLEMENTED);
+    }
+
+    struct aws_h2_stream *stream;
+    if (s_get_active_stream_for_incoming_frame(connection, stream_id, AWS_H2_FRAME_T_HEADERS, &stream)) {
+        return AWS_OP_ERR;
+    }
+
+    if (stream) {
+        if (aws_h2_stream_on_decoder_headers_begin(stream)) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int s_decoder_on_headers_i(
+    uint32_t stream_id,
+    const struct aws_http_header *header,
+    enum aws_http_header_name name_enum,
+    enum aws_http_header_block block_type,
+    void *userdata) {
+
+    struct aws_h2_connection *connection = userdata;
+    struct aws_h2_stream *stream;
+    if (s_get_active_stream_for_incoming_frame(connection, stream_id, AWS_H2_FRAME_T_HEADERS, &stream)) {
+        return AWS_OP_ERR;
+    }
+
+    if (stream) {
+        if (aws_h2_stream_on_decoder_headers_i(stream, header, name_enum, block_type)) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int s_decoder_on_headers_end(
+    uint32_t stream_id,
+    bool malformed,
+    enum aws_http_header_block block_type,
+    void *userdata) {
+
+    struct aws_h2_connection *connection = userdata;
+    struct aws_h2_stream *stream;
+    if (s_get_active_stream_for_incoming_frame(connection, stream_id, AWS_H2_FRAME_T_HEADERS, &stream)) {
+        return AWS_OP_ERR;
+    }
+
+    if (stream) {
+        if (aws_h2_stream_on_decoder_headers_end(stream, malformed, block_type)) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int s_decoder_on_end_stream(uint32_t stream_id, void *userdata) {
+    struct aws_h2_connection *connection = userdata;
+
+    /* Not calling s_get_active_stream_for_incoming_frame() here because END_STREAM
+     * isn't an actual frame type. It's a flag on DATA or HEADERS frames, and we
+     * already checked the legality of those frames in their respective callbacks. */
+
+    struct aws_hash_element *found = NULL;
+    aws_hash_table_find(&connection->thread_data.active_streams_map, (void *)(size_t)stream_id, &found);
+    if (found) {
```

Comment:
If this is not found, is that a problem? Seems like that should be impossible?

Reply:
We're just reading frames off the wire. Nothing stopping a bad implementation or attacker from sending us frames for a stream that doesn't exist